lib/semantics/resolve-names.cc

```diff
@@ -1099,7 +1207,62 @@ class OmpVisitor : public virtual DeclarationVisitor {
     return false;
   }
 
+  void Post(const parser::Name &);
+
 private:
+  struct OmpContext {
+    OmpContext(const parser::CharBlock &source, OmpDirective d)
+      : directiveSource{source}, directive{d} {}
+    parser::CharBlock directiveSource{nullptr};
+    OmpDirective directive;
+    Scope *scope{nullptr};
+    // TODO: default DSA is implicitly determined in different ways
+    Symbol::Flag defaultDSA{Symbol::Flag::OmpShared};
+    // variables on Data-sharing attribute clauses
+    std::map<const Symbol *, Symbol::Flag> objectMap;
+    bool withinConstruct{false};
+  };
+  // back() is the top of the stack
+  OmpContext &GetContext() {
+    CHECK(!ompContext_.empty());
+    return ompContext_.back();
+  }
+  void PushContext(const parser::CharBlock &source, OmpDirective dir) {
+    ompContext_.emplace_back(source, dir);
+    SetContextScope(source);
+  }
+  void PopContext() { ompContext_.pop_back(); }
+  void SetContextDirectiveSource(parser::CharBlock &dir) {
+    GetContext().directiveSource = dir;
+  }
+  void SetContextDirectiveEnum(OmpDirective dir) {
+    GetContext().directive = dir;
+  }
+  void SetContextScope(const parser::CharBlock &source) {
+    auto &currScope{context_.FindScope(source)};
+    GetContext().scope = &currScope;
+  }
+  const Scope &currScope() { return *GetContext().scope; }
+  void SetContextDefaultDSA(Symbol::Flag flag) {
+    GetContext().defaultDSA = flag;
+  }
+  void SetContextObjectMap(const Symbol *symbol, Symbol::Flag flag) {
```

Comment:
This function should take a `const Symbol &` because nullptr is never passed in.